LEP for AD refactoring

PhanLe1010 opened the pull request:
Second approach for the AD refactoring.

longhorn/longhorn#3715

cc @joshimoo for guidance.

shuo-wu commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+   1. csi-attacher send attaching request to longhorn-csi-plugin
+   1. longhorn-csi-plugin sends attaching request to longhorn-manager with pod-id and attacherType `csi-attacher`
+   > Note: maybe we can use VA name instead of pod-id
+   2. longhorn manager modify the VolumeAttachment object with this spec:
```
`modify or create`?

shuo-wu commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+            longhornvolume: <volume-name>
+      spec:
+         attachmentRequests:
+            # If the UI send multiple requests to different nodes, what should we do? Keep both or overwrite
```
IMO, we should reject the following requests.

shuo-wu commented on enhancements/20221024-longhorn-volumeattachment.md:
````diff
@@ -0,0 +1,373 @@
+      status:
+         attachmentRequests: {}
+      ```
+   1. Longhorn UI watches the `volumeAttachment.Status.AttachmentRequests[<pod-id>].Attached` and `volumeAttachment.Status.AttachmentRequests[<pod-id>]AttachError`
````
Is this a blocking call? What if the volume is attached to another node (restoring data for example)?

PhanLe1010 replied:
This is not a blocking call. 

> Is this a blocking call? What if the volume is attached to another node (restoring data for example)?

In this case Longhorn UI will display `volumeAttachment.Status.AttachmentRequests["longhorn-api"]AttachError`  with the message `volume already attached to different node`. How do you think?

PhanLe1010 replied:
Maybe the request just fails at the API layer (VA Spec update fails)

shuo-wu commented on enhancements/20221024-longhorn-volumeattachment.md:
````diff
@@ -0,0 +1,373 @@
+      ```
+   1. Backup controller wait for volume to be attached and take the backup
+1. Auto attach/detach volume for K8s CSI snapshot
+   1. csi-snappshotter send a request to longhorn-csi-plugin to with snapshot name and volume name
````
What if the csi snapshot is type backup?

PhanLe1010 replied:
Updated. Thank you. It will watch the backup CR instead

shuo-wu reviewed (commented):
There are lots of details unclear:
1. How does the AD controller handle multiple attach requests at the same time?
2. In the state illustration, why the state can be modified from `attaching` to `detaching` or in a reverse way? Why the state can be modified from `detached` to `detaching`? ...
3. More details about the orchestration between node ID in Volumes and that in VolumeAttachments.

PhanLe1010 commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+
+### Goals
+
+Introduce a new resource (Longhorn volumeattachement) to cover the following scenarios for Longhorn volume's AD:
```
Updated. Thank you

PhanLe1010 commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+type VolumeAttachmentSpec struct {
+	// maybe just name Attachment
+	AttachmentRequests map[string]AttachmentRequest `json:"attachmentRequests"`
+	// AutoSalvageNeeded bool
```
Updated. Thank you

PhanLe1010 commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+type AttachmentRequest struct {
+	// The node ID on which the controller is responsible to reconcile this orphan CR.
+	// +optional
+	ID         string            `json:"id"`
```
Updated. Thank you

PhanLe1010 commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+
+// Snapshot is the Schema for the snapshots API
+type VolumeAttachment struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
```
No since VolumeAttachment object is controlling the volume object (setting/deleting volume.NodeID and attachment parameters). The volume controller is not creating VolumeAttachment object

PhanLe1010 replied:
TODO: set the ownership reference of VA to the volume to have the cascading deletion behavior

PhanLe1010 commented on enhancements/20221024-longhorn-volumeattachment.md:
```diff
@@ -0,0 +1,373 @@
+This controller watches the VolumeAttachment objects of a volume.
+   1. Detach loop
+   Loop over VolumeAttachmentSpec.AttachmentRequests map and decide if the volume is no longer needed to be
+   attached to the current volume.Spec.NodeID. If so, reset the volume.Spec.NodeID to empty.
```
After this feature, user will not set volume.spec.NodeID directly

PhanLe1010 replied:
TODO: add a comment to explain how users use the VA

Volume is currently attached to node-1
ownerid is node-1
csi-snaphotter want to attach to node-1 (OK)
ownerid is node-2
recurring-job want to attach to node-2 (FAILED)


auto-attachment, we will choose the node as below:
1. v.Status.CurrentNodeID != "" , then we choose v.Status.CurrentNodeID
1. v.Spec.NodeID
1. v.Status.OwnerID

PhanLe1010 commented:
> How does the AD controller handle multiple attach requests at the same time?

Updated. Thank you

> In the state illustration, why the state can be modified from attaching to detaching or in a reverse way? Why the state can be modified from detached to detaching? ...

This is the hardest part of this feature. The idea is that we want to make volume controller to be resilient and can always transition to a stable state from any state. I am still thinking about this. But I think as long as there is only one outgoing arrow from each state and there is no cycle inside the graph we model it. 

The state can be modified from attaching to detaching by starting to stop the engine and the replica
From detaching to attaching: I cannot see this case in the graph
The state can be modified from detached to detaching by attempting to verify the status of the engine and replica and make sure that they are really stopped. 


> More details about the orchestration between node ID in Volumes and that in VolumeAttachments.

Updated. Thank you

innobead commented:
@PhanLe1010 Please fix the codespell check error. Thanks.

PhanLe1010 commented:
@innobead We are still changing the design as [the backend PR](https://github.com/longhorn/longhorn-manager/pull/1541) evolve. I will update the LEP once we agree on the design

innobead commented:
@PhanLe1010 Sure. Need to make sure the LEP updated.

innobead commented:
Let's update this after the PR review is ready.

innobead commented:
@PhanLe1010 Is this LEP up to date? also need to fix the PR check failure.

PhanLe1010 commented:
Sorry that I miss this PR. Yes, I will update this PR soon.

innobead commented:
@PhanLe1010 Please help update this PR to reflect the current implementation in 1.5, then we can merge it.

innobead commented:
@PhanLe1010 Let's update this and merge it, since it has been introduced already.

innobead reviewed (approved):
LGTM.

innobead commented:
cc @longhorn/dev if you still don't know about Longhorn VA, please check this PR.
